### ISSUE TITLE
drake-visualizer needs test coverage in CI

### ISSUE THREAD

**Issue by RussTedrake** [RobotLocomotion/drake#2767]
https://github.com/RobotLocomotion/drake/issues/2731 should have been caught immediately by CI -- drake-visualizer simply failed to run on homebrew mac.  We do not currently have any test coverage for running the drake-visualizer.

**Comment by david-german-tri** [comment]
If I'm understanding correctly, this issue is asking for regression tests that run the whole `drake-visualizer` executable.  To do this right, we'd need a framework that allows us to make assertions on the graphical output.  However, if we just want to test that `drake-visualizer` starts up, we could modify it to emit an "I'm ready" message.

**Comment by liangfok** [comment]
I believe that for now just verifying that `drake-visualizer` starts up would be sufficient since that would have caught #2731. Another option to having it emit an "I'm ready" message would be to add a command line argument that causes it to start up but then immediately shut down with a special exit code that can then be checked.

**Comment by wxmerkt** [comment]
@david-german-tri director has some basic tests itself already, cf. https://github.com/RobotLocomotion/director/blob/master/src/python/tests/CMakeLists.txt

I believe adding one that will start https://github.com/RobotLocomotion/director/blob/master/src/python/director/drakevisualizer.py using directorPython and running the basic director test suite as part of the drake CI coverage might do the trick (with a timer to quit after 5 seconds).

**Comment by david-german-tri** [comment]
@wxmerkt Thanks for the pointer!  Taking a look now.

**Comment by david-german-tri** [comment]
When I run `ctest -VV` in `drake-distro/externals/director/distro/pods/drake-distro/pod-build/src/director-build/src/python/tests`, most of the tests fail due to missing `import`s. The reason is that the Drake superbuild doesn't build all of Director.  We probably want to keep it that way.  So, I'm going to explore a black-box test instead.

https://github.com/RobotLocomotion/drake/blob/master/CMakeLists.txt#L167

**Comment by david-german-tri** [comment]
As an additional wrinkle, it turns out `drake-visualizer` and `directorPython` understandably will not start up without an X server.

**Comment by jwnimmer-tri** [comment]
Does http://manpages.ubuntu.com/manpages/trusty/man1/xvfb-run.1.html help?

**Comment by david-german-tri** [comment]
Interesting.  Perhaps so, if I can figure out the right incantation to turn on GLX.

**Comment by mwoehlke-kitware** [comment]
> When I run `ctest -VV` in `drake-distro/externals/director/distro/pods/drake-distro/pod-build/src/director-build/src/python/tests`, most of the tests fail due to missing `import`s. The reason is that the Drake superbuild doesn't build all of Director.

Why is director emitting tests for stuff that wasn't built?

**Comment by david-german-tri** [comment]
It looks like `xvfb-run` works just fine with GLX on AWS.  (Locally, it doesn't play nicely with my NVIDIA driver.)  So, my high-level plan is:

1. Write a Python script that runs drake-visualizer for a few seconds.  I have yet to find an approach that lets me send a graceful quit signal to the `DrakeVisualizerApp`.  Worst-case, I don't actually need to start the app, in the manner of https://github.com/RobotLocomotion/director/blob/master/src/python/tests/testDrakeVisualizer.py

2. Run that script using `directorPython` in a CMake `add_test` clause.  On platforms where a display is present, that just works.  If there is no display, we need `xvfb-run`.  That entails some subtasks:

2a. In CMake, detect whether a display is present or not.  Anyone have ideas about how to do that?
2b. Add Xvfb to the Linux CI images.
2c. Confirm there's a usable display in the Mac CI context.

(There will be no coverage for Windows because Drake doesn't yet build director on Windows.)

**Comment by david-german-tri** [comment]
> Why is director emitting tests for stuff that wasn't built?

If I'm understanding it correctly, the directory containing the tests is built, but not all of the dependencies for all of the python code under test are built.  `import` errors ensue.

**Comment by jwnimmer-tri** [comment]
> On platforms where a display is present, that just works. If there is no display, we need xvfb-run. That entails some subtasks:
> 2a. In CMake, detect whether a display is present or not. Anyone have ideas about how to do that?

Perhaps: in the test, if there is no X, then `subprocess.check_call()` ourself recursively with xvfb-run stuck first.  That way the user experience is just run the test, and it sets up stubs as needed.

**Comment by mwoehlke-kitware** [comment]
> the directory containing the tests is built, but not all of the dependencies for all of the python code under test are built

Let me rephrase: is there a reason why director calls `add_test` for tests with optional dependencies without checking if the dependencies are enabled?

**Comment by mwoehlke-kitware** [comment]
> In CMake, detect whether a display is present or not. Anyone have ideas about how to do that?

Check if `$DISPLAY` is set?

TBH, I'd consider just always using a VFB, to avoid obnoxiously popping up windows. Maybe as an option (default=ON) in case of systems where xvfb doesn't work for whatever reason.

**Comment by rdeits** [comment]
Pat reports successfully using xvfb with director in the past: #1256
